### DISCUSSION TITLE
USD curves in Maya 2022?

### DISCUSSION THREAD

**Discussion by BrianHanke** [Autodesk/maya-usd#1397] [Q&A]
I've got a head created in Houdini with a full set of hair stored as curves. I exported from Solaris and I can open the USD file fine in usdview and Clarisse. Maya loads the geometry but not the curves. They show up as empty items in the outliner with notes like:

`Imported from @c:/Users/Brian/Desktop/obj/hair_l/material1.usd@</hair/hair_l/purple01> with type 'BasisCurves'`

Is there a way to get USD curves working in Maya? Thanks!

**Comment by williamkrick** [comment]
Hi Brian, we do have some known issues right now with basis curves that can prevent them from showing up.

If your viewport display mode is "textured" or you have "Use Default Material" checked under the Shading options then the basis curves will not be displayed. There are also some issues with material bindings on basis curves.

If that doesn't solve your problem can you share a simplified usda file which shows the issue? A single curve would be great.

**Comment by BrianHanke** [comment]
Thanks for the reply, William. I created a single curve in Houdini and exported to USD. I've attached the USD file and pics of my results in different apps. The original Houdini, the curve displaying correctly in usdview, the curve not displaying as a simple Maya import (although polygon geo does display fine when done this way) and the curve displaying correctly when used to create a USD stage. So, it's half working in Maya, but the stage context seems to be very locked down in terms of interaction and rendering and my goal is to be able to import USD from Houdini and have all the geo accessible for shading and rendering in Arnold.

![curve03](https://user-images.githubusercontent.com/59420805/117314396-b7dafc80-ae54-11eb-9dc1-112b63ecb7f5.png)
![curve04](https://user-images.githubusercontent.com/59420805/117314416-bdd0dd80-ae54-11eb-9af4-38c7c65f9e53.png)
![curve01](https://user-images.githubusercontent.com/59420805/117314429-c1646480-ae54-11eb-9a66-db4cf82cece0.png)
![curve02](https://user-images.githubusercontent.com/59420805/117314438-c45f5500-ae54-11eb-8e50-c379fc0ed21b.png)
[curve.zip](https://github.com/Autodesk/maya-usd/files/6435270/curve.zip)

**Reply by williamkrick** [reply]
I am able to reproduce the curve working correctly when it is brought in as USD directly in the proxy shape and I am able to reproduce the curve not displaying as an imported NURBS curve. It seems that the curve shape is not created at all on import. I would consider the curve failing to import a defect. I logged internal defect MAYA-111503 to track this one.

You mentioned that you found the stage context to be very locked down. Can you talk a little bit about the types of things you'd like to be able to do to native USD data that you can't right now?

**Reply by BrianHanke** [reply]
Thanks for flagging it as a defect!

As far as the current Maya USD workflow I'd say the main problem is that it doesn't behave like Houdini or Clarisse in terms of exposing everything in the scene to editing and material assignment. I think the standard import would be fine if curves were working. Creating a stage from a file appears to leave very little exposed to the user. For example, in this scene both Clarisse and Solaris give full access to everything. You can assign materials and the lights and cameras are imported. 

Bringing the same thing into Maya via File > Import works as expected for the most part. The cameras are there and you can assign Arnold materials to poly geo. Really the only issue in my case is the curve thing. If that was fixed I think I'd be very satisfied with the  experience. I'm mostly looking for a way to bring in heavy scenes in a quick and easy way. Trying to import all that hair as Alembic or FBX is extremely slow, but USD takes no time at all.

On the other hand, import to stage seems to flatten the USD file, no cameras are there and none of the layers are listed. Although... layers are selectable in the viewport, so they're sort of half there. And as far as I can tell you can't assign any Arnold materials or render it. If I mess around with the numerous USD export options in Solaris I can get the layers to show up in the Maya stage, but that isn't required in Clarisse or usdview.

![houdini](https://user-images.githubusercontent.com/59420805/117323903-4e132080-ae5d-11eb-979a-0a81ea79f816.png)
![clarisse](https://user-images.githubusercontent.com/59420805/117323913-510e1100-ae5d-11eb-973a-36070235aadb.png)
![maya01](https://user-images.githubusercontent.com/59420805/117323927-53706b00-ae5d-11eb-94b6-fe7469d16399.png)
![maya02](https://user-images.githubusercontent.com/59420805/117323940-579c8880-ae5d-11eb-9ea3-53a45b71e3ec.png)

**Comment by wtelford** [comment]
Hey Brian,

There are two ways to bring USD in to Maya
- Import
- ProxyShape

The import translates all USD data to Maya data. From that point forward you are just working with Maya data. There are many translators to handle different data types, but not every known data type has a known translator. You ran in to a bug with our curve translator that William logged above.

The ProxyShape keeps a complete USD stage in memory without any translation. This supports everything USD can handle as it is USD. When using the proxyShape you should display "shapes" in the Maya outliner, as all of the prims in the scenegraph are shown under the proxyShape. In the screenshots above you are only seeing the Maya transform that is the parent of the USD proxyShape. All prims in the USD file should be visible in the outliner. I believe this is why you feel the USD data had been "flattened". The viewport (VP2)  will also draw many but not yet all prims. An example of a prim that is not yet supported in the viewport would be the "points" schema. 

The layer editor will show you the structure of the root sub-layer stack. We currently do not have a viewer for other composition arcs (references, payloads, variants, inherits, specializes), but references can be authored from a right-click menu in the Outliner. 

We also do not yet have a material authoring workflow as our focus to this point has been layout & assembly.

You will find that through the Maya script editor you do have complete access to query and modify the USD stages in memory with the Pixar USD python API.

**Reply by BrianHanke** [reply]
@wtelford Thanks for the breakdown! USD is a whole can of worms above my pay grade, but I really like it as a scene transfer format. For my purposes once the curves bug is fixed and it understands curves as Maya data I'll be a happy camper. I do feel that my final screenshot shows a potential issue as those particular scene layers are not showing up in either the outliner or the Layer Editor. They're there in Houdini, usdview and Clarisse. But, I may be missing something. Advanced USD is definitely not my strength! I might prepare a simpler scene and share here to ask about what I'm doing wrong.

**Reply by wtelford** [reply]
Hey Brian,
Can you clarify what you mean by layers? If you're using Maya 2022, make sure you open Maya's attribute editor when you select USD prims in the the viewport or the outliner. You will see all of the properties on the prim as well.

**Reply by BrianHanke** [reply]
@wtelford It's hard to describe, so here are a few quick videos. This scene was created in Houdini, staged and exported to USD in Solaris. Video 1 is loading the scene in usdview. Everything is there as expected. Video 2 is loading in Maya via File > Import. All the scene contents are in the outliner, but of course the curves aren't showing and Arnold (Hydra) doesn't pick up any of the baked-in colors. Finally, video 3 shows the scene imported to a stage. There is nothing in the outliner and nothing in the USD layer editor. However, each prim is selectable in the viewport. Shouldn't they be visible in the outliner and especially the layer editor? Please let me know if I'm missing something. Thanks!

https://user-images.githubusercontent.com/59420805/117384506-23eb4e00-aeb1-11eb-916a-600303944f5c.mov

https://user-images.githubusercontent.com/59420805/117384532-35ccf100-aeb1-11eb-8cd4-0ce6bacc79aa.mov

https://user-images.githubusercontent.com/59420805/117385276-c526d400-aeb2-11eb-8821-e245993b4a41.mov

**Reply by wtelford** [reply]
Brian,
For video 3 please turn on display shapes in the outliner. The entire scene graph hierarchy will display.

**Reply by BrianHanke** [reply]
@wtelford Aha, there it all is! You said there's no way to add Arnold shaders yet, right? The Karma materials are there but of course they don't render. I still need to import via the file menu if I want geometry that Maya recognizes?

**Comment by wtelford** [comment]
We currently do not yet provide material authoring/assignment workflows directly on the USD data model. 
If you are working with a Maya data model materials and material assignments can be exported to USD.

**Reply by BrianHanke** [reply]
Understood. Thanks a lot for all your help!

**Reply by wtelford** [reply]
No problem Brian. The following links might help as well for anyone else who may stumble across this topic in the future.
[Getting Started](https://help.autodesk.com/view/MAYAUL/2022/ENU/?guid=GUID-36CFE2C3-766F-4B00-8464-E94F95E7AF4B)
 [USD in the Maya Interface](https://help.autodesk.com/view/MAYAUL/2022/ENU/?guid=GUID-2FFFAB8A-022B-42BF-9068-F5BAD45238A9)
[USD in the Maya Outliner](https://help.autodesk.com/view/MAYAUL/2022/ENU/?guid=GUID-97358ED4-179F-4568-9382-7409E3F8E70A)
[USD Layer Editor](https://help.autodesk.com/view/MAYAUL/2022/ENU/?guid=GUID-4FAD73CA-E775-4009-9DCB-3BC6792C465E)

**Comment by BrianHanke** [comment]
@wtelford @williamkrick I noticed that the latest changelog has an item about basis curves. I did a few tests and the behavior seems the same as before, no curves displaying when importing USD via the file menu. Just checking to see if I'm missing something about this update or if I should go back into hibernation mode until the next round of changes. Thanks!

**Reply by williamkrick** [reply]
Hi Brian,

Fixing the basis curve import to NURBS has been logged internally as MAYA-111503. Unfortunately the work to fix importing basis curves hasn't been scheduled yet. The work in 0.10.0 was to add a test to prevent the situation from getting any worse.

Thanks for keeping a close eye on it!

**Reply by BrianHanke** [reply]
Got it, thanks for the reply. Back into hibernation!
